qiskit/qasm3/__init__.py

```diff
@@ -365,11 +396,35 @@ def loads(
                 + qiskit_qasm3_import.__version__
             )
         kwargs["annotation_handlers"] = annotation_handlers
+
     try:
-        return qiskit_qasm3_import.parse(program, **kwargs)
+        qasm3_ckt = qiskit_qasm3_import.parse(program, **kwargs)
     except qiskit_qasm3_import.ConversionError as exc:
         raise QASM3ImporterError(str(exc)) from exc
 
+    if num_qubits is not None:
+        if qasm3_ckt.num_qubits > num_qubits:
+            raise ValueError(
+                "Number of qubits cannot be more than the provided number of physical/virtual qubits."
+            )
+        if (difference := num_qubits - qasm3_ckt.num_qubits) > 0:
+            if qasm3_ckt.layout is not None:
+                # updating layout
+                last = max(qasm3_ckt.layout.initial_layout.get_physical_bits().keys(), default=-1)
+                for k in range(last + 1, num_qubits):
+                    qasm3_ckt.layout.initial_layout.add(Qubit(), k)
+
+                # updating input_qubit_mapping
+                mapping_dict = {}
+                for ii in range(last + 1, num_qubits):
+                    mapping_dict[qasm3_ckt.layout.initial_layout[ii]] = ii
+                qasm3_ckt.layout.input_qubit_mapping.update(mapping_dict)
```

Comment:
Let's take the layout _out_ of the circuit (`layout, qasm3_ckt._layout = qasm3_ckt._layout, None`), mutate it, add the bits to the circuit, _and then_ put the layout back on the circuit, just so we never call `add_bits` while the layout is actually set.  Given #14572, I think that probably makes more sense.